### PR TITLE
Update mutable state for nDC compatibility

### PR DESCRIPTION
**Pull request by wxing1292**
* Unify mutable state current version, start version and last write version APIs
* Unify mutable state set / get branch token APIs
* Add GetCurrentVersionHistory in version history for ease of use
* Rename branch index to version history index
* Modify mutable state get start / completion events APIs to return err, instead of bool
* If version history is set, then branch token should be accessed using version history, otherwise, using execution info
* Mutable state next event ID is kept as is, within execution info

### CODE REVIEW

File: service/history/mutableStateBuilder.go [hunk 1]
```diff
@@ -91,10 +106,13 @@ type (
...
 		// in memory only attributes
+		// indicate the current version
+		currentVersion int64
 		// indicates whether there are buffered events in persistence
 		hasBufferedEventsInDB bool
-		// indicates whether the workflow is pointed by current record, guaranteed
-		isWorkflowBeingCurrentGuaranteed bool
+		// indicates the workflow state in DB, can be used to calculate
+		// whether this workflow is pointed by current workflow record
+		stateInDB int
```
Comment (meiliang86) [inline service/history/mutableStateBuilder.go]: Do we really want to emphasis "InDB"?

File: service/history/mutableStateBuilder.go [hunk 2]
```diff
@@ -621,14 +732,30 @@ func (e *mutableStateBuilder) assignTaskIDToEvents() error {
 }
 
 func (e *mutableStateBuilder) IsCurrentWorkflowGuaranteed() bool {
```
Comment (meiliang86) [inline service/history/mutableStateBuilder.go]: Not clear what "Guaranteed" means in this context. I can sort of guess based on the code. Might worth putting comments on this.

File: service/history/mutableStateBuilder.go [hunk 3]
```diff
@@ -3804,6 +4143,30 @@ func (e *mutableStateBuilder) syncActivityToReplicationTask(
 	)
 }
 
+func (e *mutableStateBuilder) updateWithLastWriteEvent(
+	lastEvent *workflow.HistoryEvent,
+) error {
+
+	e.GetExecutionInfo().LastEventTaskID = lastEvent.GetTaskId()
+
+	if e.replicationState != nil {
```
Comment (meiliang86) [inline service/history/mutableStateBuilder.go]: Move nil check inside UpdateReplicationStateLastEventID?

File: common/persistence/versionHistory.go [hunk 4]
```diff
@@ -520,18 +528,30 @@ func (h *VersionHistories) IsRebuilt() (bool, error) {
...
-	h.currentBranchIndex = index
+	h.currentVersionHistoryIndex = index
 	return nil
 }
 
-// GetCurrentBranchIndex get the current branch index
-func (h *VersionHistories) GetCurrentBranchIndex() int {
-	return h.currentBranchIndex
+// GetCurrentVersionHistoryIndex get the current branch index
+func (h *VersionHistories) GetCurrentVersionHistoryIndex() int {
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/persistence/versionHistory.go]: i will probably write a readme about nDC, or just expose the design doc

[the remark this answers is not in the document]
Comment (wxing1292) [reply common/persistence/versionHistory.go]: #2290

File: common/persistence/versionHistory.go [hunk 5]
```diff
@@ -181,15 +181,23 @@ func (v *VersionHistory) DuplicateUntilLCAItem(
 	return nil, notFoundErr
 }
 
-// SetBranchToken the set the branch token
-func (v *VersionHistory) SetBranchToken(
+// InitializeBranchToken the set the branch token
+func (v *VersionHistory) InitializeBranchToken(
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/persistence/versionHistory.go]: deleted

File: common/persistence/versionHistory.go [hunk 6]
```diff
@@ -181,15 +181,23 @@ func (v *VersionHistory) DuplicateUntilLCAItem(
...
 
 	if len(v.branchToken) != 0 {
 		return &shared.BadRequestError{Message: "branch token is already set"}
 	}
 
+	return v.SetBranchToken(inputToken)
+}
+
+// SetBranchToken the set the branch token
+func (v *VersionHistory) SetBranchToken(
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/persistence/versionHistory.go]: i would keep the return error for now, for ease of use in the future

File: service/history/decisionHandler.go [hunk 7]
```diff
@@ -117,9 +117,9 @@ func (handler *decisionHandlerImpl) handleDecisionTaskScheduled(
 				transferTasks:  []persistence.Task{&persistence.RecordWorkflowStartedTask{}},
 			}
 
-			startEvent, found := msBuilder.GetStartEvent()
-			if !found {
-				return nil, &workflow.InternalServiceError{Message: "Failed to load start event."}
+			startEvent, err := msBuilder.GetStartEvent()
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/decisionHandler.go]: this is not tied to decision task or any other task.

if a workflow cannot access the start event, then this workflow is corrupted.

File: service/history/mutableStateBuilder.go (hunk 1 above)
Comment (wxing1292) [reply service/history/mutableStateBuilder.go]: yes, nextEventIDInDB is used for compare and swap

stateInDB is used like a bloom filter, i.e. stateInDB being created / running meaning that this workflow must be the current workflow (assuming there is no rebuild of mutable state).

stateInDB being completed does not guarantee this workflow being the current workflow
stateInDB being zombie guarantees this workflow not being the current workflow

File: service/history/mutableStateBuilder.go (hunk 3 above)
Comment (wxing1292) [reply service/history/mutableStateBuilder.go]: this is to avoid breaking all kinds of tests......

**Comment by wxing1292** [comment]
> Consider moving verifyTaskVersion into queue processor base? 

the thing is, sometimes, we need to verify the task version, sometimes, we cannot verify the task version, i.e. it depends.

**Pull request merged**